### ISSUE TITLE
benchdnn:  symbol lookup error 

### ISSUE THREAD

**Issue by vitduck** [oneapi-src/oneDNN#1072]
# Summary
benchdnn failed with following error message: 
undefined symbol: _ZN4dnnl4impl14is_pd_in_cacheEPK19dnnl_primitive_desc

# Version
2.3 

# Environment
Intel DevCloud 
* GPUs: Gen9 and Iris_Xe
* CMake version 3.16.3
* gcc 9

# build oneDNN from source (success)
export CC=icc 
export CXX=dpcpp 
cmake .. -DDNNL_CPU_RUNTIME=DPCPP -DDNNL_GPU_RUNTIME=DPCPP 

# from the documentation of benchdnn: 
./benchdnn --engine=gpu:1--concat --batch=inputs/concat/test_concat_all
./benchdnn: symbol lookup error: ./benchdnn: undefined symbol: _ZN4dnnl4impl14is_pd_in_cacheEPK19dnnl_primitive_desc

Above is the error I have enountered with oneDNN build from source on Intel DevCloud. 
I have tried different engines such as --conv, --matmul and got the aformentioned error.
I am puzzled because only some of the tests failed, yet the benchmarks did not run at all. 
Thanks

**Comment by vpirogov** [comment]
Thank you for the report, @vitduck.

It looks like benchdnn picks up older oneDNN library installed in the system, instead of the one from your local build. You can test whether this is the case by running `ldd benchdnn` command.

**Comment by vitduck** [comment]
Yes. This is indeed an oversight on my part. I fogot that oneAPI has its own oneDNN library. 
` libdnnl.so.2 => /glob/development-tools/versions/oneapi/2021.2/inteloneapi/dnnl/2021.2.0/cpu_dpcpp_gpu_dpcpp/lib/libdnnl.so.2 (0x00007f4332a39000)
        libsycl.so.5 => /glob/development-tools/versions/oneapi/2021.2/inteloneapi/compiler/2021.2.0/linux/lib/libsycl.so.5 (0x00007f43327af000)`
Using LD_PRELOAD does the the trick. 
Thank you for your time.

**Comment by vpirogov** [comment]
@vitduck, 

I can reproduce the segfault on CPU when oneDNN is built with a combination of `icc` and `dpcpp` compiler drivers. This combination of drivers is inconsistent and is not supported by oneDNN. Everything works as expected with `clang` and `clang++`, as recommended in [oneDNN developer guide](https://oneapi-src.github.io/oneDNN/dev_guide_build.html).

**Comment by vpirogov** [comment]
The message is somewhat cryptic, but essentially it says that it cannot read the problem descriptor. 

The reason you see it is typo in `-v 1` option. The following line works:
```
./benchdnn --engine=cpu --mode=P --conv -v 1 --batch=inputs/conv/perf_conv_clx_1sock
```

**Comment by vitduck** [comment]
Thanks. 

I was using the verbose option to see why a segmentation fault has occured with performance mode (--mode=p): 
`./benchdnn --engine=cpu --mode=P --conv -v99 --batch=inputs/conv/perf_conv_clx_1sock`
`batch file used: inputs/conv/set_perf_cpu_large_mb
batch file used: inputs/conv/shapes_1d_wavenet
run: --conv --mode=P --dir=FWD_D g1mb1ic32iw84350oc32ow84349kw2pw0n"wavenet:conv1d_1"
oneDNN implementation: jit:avx512_common
oneDNN implementation: jit:avx512_common
Requested: 0.0402362 GB, benchdnn limit: 23.288 GB, CPU RAM capacity: 31.0506 GB, GPU RAM capacity: 0 GB
Segmentation fault
` 

Below is the output for 'wavenet:conv1d_1' with correctness mode (default): 
 `./benchdnn --conv --dir=FWD_D -v6 g1mb1ic32iw84350oc32ow84349kw2pw0n"wavenet:conv1d_1"`
`run: --conv --dir=FWD_D g1mb1ic32iw84350oc32ow84349kw2pw0nwavenet:conv1d_1
oneDNN implementation: jit:avx512_common
oneDNN implementation: jit:avx512_common
Requested: 0.0402362 GB, benchdnn limit: 23.288 GB, CPU RAM capacity: 31.0506 GB, GPU RAM capacity: 0 GB
@@@ [DST] final: diff: err:0, l0(``-0``) l1:(2.4669e+09,2.4669e+09,0,``0``) l2:(1.83226e+06,1.83226e+06,0,``0``) l8:(4048,4048,-0,``-0``)
0:PASSED __REPRO: --conv --dir=FWD_D g1mb1ic32iw84350oc32ow84349kw2pw0nwavenet:conv1d_1
tests:1 passed:1 skipped:0 mistrusted:0 unimplemented:0 failed:0 listed:0
`

I think the segmentation fault only occurs in performance mode.

**Comment by vpirogov** [comment]
@vitduck, thank you for the report. I confirm segfault in benchdnn with performance mode enabled. The segfault goes away with `DNNL_VERBOSE=1` or in correctness mode, so this may be an issue with benchmarking driver. We'll investigate.

**Comment by vitduck** [comment]
For the moement, I am able to obtain performance on cpu engine with `DNNL_VERBOSE=1`. 
```
perf,cpu,jit:avx2,"wavenet:conv1d_1",--conv --mode=P --dir=FWD_D g1mb1ic32iw84350oc32ow84349kw2pw0n"wavenet:conv1d_1",0.345494,0,3.32544,103.894,3.46721,99.6461
```
Thanks for your time. I've learned alot through the discussion.

**Comment by densamoilov** [comment]
The root cause of the issue is a missing `wait()` call after submitting CPU primitive to the queue. The issue is fixed in c79bfd5b8048e606650d2eee8f6818efdeb9215e. Thanks for reporting the issue.

**Comment by vpirogov** [comment]
The fix was promoted to oneDNN v2.3 as well.

**Comment by isuruf** [comment]
I ran into the same error with a
```
./tests/benchdnn/benchdnn: symbol lookup error: ./tests/benchdnn/benchdnn: undefined symbol: _ZN4dnnl4impl3cpu8platform20has_training_supportE16dnnl_data_type_t
```
because the Intel devcloud machines are activated with `LD_LIBRARY_PATH` which points to an older oneDNN library.

The fix is to add `-Wl,--disable-new-dtags` to make sure `RPATH` is set which has higher precedence than `LD_LIBRARY_PATH`. (Otherwise `RUNPATH` is set which has lower precedence than `LD_LIBRARY_PATH`). Of course `LD_PRELOAD` should work too.

It would be good to add `-Wl,--disable-new-dtags` to benchdnn flags.
